File: docs/exPuzzlers/src/_06_classy/_47_Well_Dog_My_Cats/Ruckus.java
package _06_classy._47_Well_Dog_My_Cats;

class Counter {
    private static int count = 0;
    public static final synchronized void increment() {
        count++;
    }
    public static final synchronized int getCount() {
        return count; 
    } 
}

class Dog extends Counter {
    public Dog() { }
    public void woof() { increment(); }
} 

class Cat extends Counter {
    public Cat() { } 
    public void meow() { increment(); }
}

public class Ruckus {
    public static void main(String[] args) { 
        Dog dogs[] = { new Dog(), new Dog() };
        for (int i = 0; i < dogs.length; i++)
            dogs[i].woof();
        Cat cats[] = { new Cat(), new Cat(), new Cat() };
        for (int i = 0; i < cats.length; i++)
            cats[i].meow();
        System.out.print(Dog.getCount() + " woofs and ");
        System.out.println(Cat.getCount() + " meows");
    }
}

File: docs/exPuzzlers/src/_06_classy/_47_Well_Dog_My_Cats/Ruckus.kt
namespace well.dog.my.cats

import std.io.*
import std.*

open class Counter {
    class object {
        private var count = 0
        fun increment() { count++ }
        fun getCount() = count
    }
}

class Dog() : Counter {
// Class objects are not inherited:
//    fun woof() { increment() }
    fun woof() { Counter.increment() }
}

class Cat() : Counter {
// Class objects are not inherited:
//    fun meow() { increment() }
    fun meow() { Counter.increment() }
}

fun main(args : Array<String>) {
    val dogs = array(Dog(), Dog())
    for (dog in dogs)
      dog.woof()
    val cats = array(Cat(), Cat())
    for (cat in cats)
      cat.meow()
// Problematic code does not compile:
//    println("${Dog.getCount()} woofs")
    println("${Counter.getCount()} woofs")
//    println("${Cat.getCount()} meows")
    println("${Counter.getCount()} meows")
}